compiler/util/src/org/jetbrains/kotlin/utils/SmartIdentityTable.kt

```diff
@@ -27,6 +27,27 @@ class SmartIdentityTable<K, V> {
     val size: Int
         get() = keysArray?.size ?: largeMap!!.size
 
+
+    val keys: Sequence<K>
+        get() = keysArray?.asSequence() ?: largeMap!!.keys.asSequence()
+
+
+    val entries: Sequence<Entry<K, V>>
```

Comment:
I'm somewhat concerned that this method behaves slightly different depending on the size of the table. Namely, if there are no more than 10 elements, it returns a new collection, otherwise it returns a view into the current collection. Which means that if you save the result of `entries` into a variable `x` and then add more elements to the table, you'll observe the newly added elements in `x` if and only if the table size was > 10 at the point where `x` was defined. This could lead to obscure bugs in certain situations. Note that `keys` doesn't suffer from this problem, it always returns a view.

I suggest either providing a view even if the table is small (more preferably, for consistency with `keys`), or return a new collection even if it's large. In any case, it might be worth to document the specific behavior (keys vs new collection) both on `keys` and `entries`